expan/core/experiment.py

```diff
@@ -1,280 +1,232 @@
-        derived_kpi_formulas = [k['formula'] for k in derived_kpis]
-
-        # what columns do we expect to find in the data frame?
-        required_column_names = (report_kpi_names_needed | experiment_column_names) - set(derived_kpi_names)
-        kpi_name_pattern = '([a-zA-Z][0-9a-zA-Z_]*)'
-        # add names from all formulas
-        for formula in derived_kpi_formulas:
-            names = re.findall(kpi_name_pattern, formula)
-            required_column_names = required_column_names | set(names)
-
-        for c in required_column_names:
-            if c not in data:
-                raise ValueError('No column %s provided'%c)
-
-        self.data                 =     data.copy()
-        self.metadata             = metadata.copy()
-        self.report_kpi_names     = report_kpi_names_needed
-        self.derived_kpis         = derived_kpis
-        self.variant_names        = set(self.data.variant)
-        self.control_variant_name = control_variant_name
-        self.reference_kpis       = {}
-
-        # add derived KPIs to the data frame
-        for name, formula in zip(derived_kpi_names, derived_kpi_formulas):
-            self.data.loc[:, name] = eval(re.sub(kpi_name_pattern, r'self.data.\1.astype(float)', formula))
-            self.reference_kpis[name] = re.sub(kpi_name_pattern + '/', '', formula)
-
-    def get_kpi_by_name_and_variant(self, data, name, variant):
-        return data.reset_index().set_index('variant').loc[variant, name]
+    """ Class which adds the analysis functions to experimental data. """
+    def __init__(self, data, metadata):
+        self.data         = data.copy()
+        self.metadata     = metadata.copy()
+        self.worker_table = {
+            'fixed_horizon': statx.make_delta,
+            'group_sequential': es.make_group_sequential,
+            'bayes_factor': es.make_bayes_factor,
+            'bayes_precision': es.make_bayes_precision
+        }
+
 
     def __str__(self):
-        variants = self.variant_names
+        return 'Experiment "{:s}" with {:d} entities.'.format(self.metadata['experiment'], len(self.data))
+
+
+    def analyze_statistical_test(self, test, testmethod, **worker_args):
+        """ Runs delta analysis on one statistical test and returns statistical results.
+        
+        :param test: a statistical test to run
+        :type  test: StatisticalTest
+        :param testmethod: analysis method
+        :type  testmethod: str
+        :param **worker_args: additional arguments for the analysis method
+        :return: statistical result of the test
+        :rtype: StatisticalTestResult
+        """
+        if not isinstance(test, StatisticalTest):
+            raise TypeError("Statistical test should be of type StatisticalTest.")
+
+        if 'entity' not in self.data.columns():
+            raise RuntimeError("There is no 'entity' column in the data.")
+        if self.data.entity.duplicated().any():
```

Comment:
I think that it used to be a concern of Expan to filter out those duplicates. Am I wrong?

Reply:
yes it is used to filter out duplicates. Should we still do the same here?

Reply:
I think that it's quite often the case that data has duplicates — I think we shouldn't break things that we don't have to break and keep at least a resemblance of backward compatibility in terms of functionality.

Reply:
oh I think I misunderstood.

Here ```entity``` should not have duplicates because we do our analysis based on this entity level. If there is duplicates, something is wrong when we group the data. And I think we should raise an error if this happens.